lib/instrumentation/modules/nodemailer.js

```diff
@@ -0,0 +1,69 @@
+var shimmer = require('../shimmer')
+
+module.exports = function (nodemailer, agent, version, enabled) {
+  if (!enabled) return nodemailer
+  agent.logger.debug('shimming nodemailer.createTransport')
+  shimmer.wrap(nodemailer, 'createTransport', function (orig) {
+    return function () {
+      var transport = orig.apply(this, arguments)
+
+      shimTransport(transport)
+      return transport
+    }
+  })
+
+  function shimTransport (transport) {
+    agent.logger.debug('shimming transport.sendMail')
+
+    shimmer.wrap(transport, 'sendMail', function (orig) {
+      return function () {
+        const transaction = agent.currentTransaction ||
+                                        agent.startTransaction('Nodemailer', 'Nodemailer')
+
+        const span = transaction.startSpan('Send Email', 'nodemailer.sendMail')
+        if (!span) {
+          return orig.apply(this, arguments)
+        }
+
+        function endTransaction (transaction) {
+          if (transaction.type === 'Nodemailer') {
+            transaction.end()
+          }
+        }
+
+        // callback is provided
+        if (typeof arguments[1] === 'function') {
+          const cb = arguments[1]
+
+          return orig.call(this, arguments[0], function (err, result) {
+            if (err) {
+              agent.captureError(err)
+            }
+            span.end()
+            endTransaction(transaction)
+
+            cb(err, result)
+          })
+        }
+
+        return new Promise((resolve, reject) => {
+          return orig.apply(this, arguments).then((result) => {
+            span.end()
+            endTransaction(transaction)
+
+            resolve(result)
+          }, (err) => {
+            agent.captureError(err)
+
+            span.end()
+            endTransaction(transaction)
+
+            reject(err)
+          })
+        })
```

Comment:
Rather than wrapping this in a new promise and having all that code duplication between success and failure handlers and the callback form, I'd just do something like this:

```js
function after (err) {
  if (err) agent.captureError(err)
  span.end()
}

// Handle cleanup for callback-form
if (typeof arguments[1] === 'function') {
  shimmer.wrap(arguments, 1, cb => {
    return function (err) {
      after(err)
      return cb.apply(this, arguments)
    }
  })
}

var result = orig.apply(this, arguments)

// Handle cleanup for promise-form
if (result && typeof result.then === 'function') {
  result.then(
    () => after(),
    after
  )
}

return result
```

Reply:
Thanks! I will update today.

> I'm currently working on an interface to allow custom instrumentation patches to be sideloaded easily

That's great! I am developing also elastic apm for meteor and experienced issue when I was not able to easily integrate my agent with yours, it forced me to hack your agent as well:
https://github.com/kschingiz/meteor-elastic-apm/blob/master/meteor-elastic-apm.js#L18

When are you going to merge PR #803? I can block my PR and wait for new release.

Reply:
It should be released very soon, I think. The PR is just waiting for review at the moment. If it all looks good, we might be able to land it early next week. :)

Reply:
@Qard Great! can we then close my PR? If you are not planning to add emailers, it should be as a separate module, I think